go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+	}()
+
+	klog.Infof("Started health check controller")
+	return nil
+}
+
+func (c *controller) run() {
+	for {
+		// pop the next probe task to execute
+		task := c.probeTaskQueue.Pop(c.ctx)
+		select {
+		case <-c.ctx.Done():
+			return
+		default:
+			// execute on its own goroutine
+			go c.probe(task)
+		}
+	}
+}
+
+func (c *controller) probe(task *probeTask) {
+	// check if task was cancelled before probing
+	nodeInfo, exists := c.isProbedWithNodeInfo(task.node)
+	if !exists || task.cancelled() {
+		task.client.Disconnect()
+		return
+	}
+
+	// probe
+	klog.V(5).Infof("Probing health state of node %s", task.node)
+	state := AVAILABLE
+	reachable := IsReachable(task.ctx, task.node, nodeInfo.ips, task.client, c.port, period)
+	if !reachable {
+		klog.Errorf("Failed health state probe for node %s", task.node)
+		state = UNREACHABLE
+	}
+
+	// update state & retask
+	updated := c.setNodeStateAndRetask(task, state)
+	if updated {
+		c.informConsumers(task.node)
+	}
+
+	// if no interest from consumers, reconcile to cancel the probing task
+	if !c.hasInterestedConsumers(task.node, nil) {
+		c.controller.Reconcile(task.node)
+	}
+}
+
+func (c *controller) reconcile(nodeName string) error {
+	start := time.Now()
+	klog.V(5).Infof("Reconciling health check for node %s", nodeName)
+	defer func() {
+		klog.V(5).Infof("Finished reconciling health check for node %s: %v", nodeName, time.Since(start))
+	}()
+
+	node, err := c.nodeInformer.Lister().Get(nodeName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if node == nil {
+		// node deleted, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node deleted", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	names := []string{}
+	if !c.hasInterestedConsumers(nodeName, &names) {
+		// no interest, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: no interested consumers out of %q", nodeName, names)
+			// in case someone just registered
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeSubnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		return fmt.Errorf("failed to parse node %s subnets annotation %v", node.Name, err)
+	}
+
+	mgmtIPs := make([]net.IP, len(nodeSubnets))
+	for i, subnet := range nodeSubnets {
+		mgmtIPs[i] = util.GetNodeManagementIfAddr(subnet).IP
+	}
+
+	update := true
+	nodeInfo, probed := c.isProbedWithNodeInfo(nodeName)
+	if probed {
+		less := func(a, b net.IP) bool { return a.String() < b.String() }
+		update = !cmp.Equal(nodeInfo.ips, mgmtIPs, cmpopts.SortSlices(less))
+	}
+
+	if !update {
+		// already probing and no config update, bail out
+		return nil
+	}
+
+	if len(mgmtIPs) == 0 {
+		// no proper config, stop probing and assume unreachable out of
```

Comment:
should we emit an error log or warning log here?

Reply:
So there is an Info log there. Would you rather change it to warning or error?

Reply:
It seems like it should at least be a warning. I don't think we expect mp0 to be missing or have no IPs. wdyt?